main/lsp/request_dispatch.cc

```diff
@@ -21,8 +25,262 @@ class SorbetFenceTask final : public LSPTask {
+    const auto it = overriddingStateHashes.find(id);
+    if (it == overriddingStateHashes.end()) {
+        return globalStateHashes[id];
+    }
+    return it->second;
+}
+
+bool canTakeFastPath(
+    const core::GlobalState &gs, const LSPConfiguration &config, const vector<core::FileHash> &globalStateHashes,
+    const LSPFileUpdates &updates,
+    const UnorderedMap<int, core::FileHash> &overriddingStateHashes = UnorderedMap<int, core::FileHash>()) {
+    Timer timeit(config.logger, "fast_path_decision");
+    auto &logger = *config.logger;
+    if (config.disableFastPath) {
+        logger.debug("Taking slow path because fast path is disabled.");
+        prodCategoryCounterInc("lsp.slow_path_reason", "fast_path_disabled");
+        return false;
+    }
+    // Path taken after the first time an update has been encountered. Hack since we can't roll back new files just yet.
+    if (updates.hasNewFiles) {
+        logger.debug("Taking slow path because update has a new file");
+        prodCategoryCounterInc("lsp.slow_path_reason", "new_file");
+        return false;
+    }
+    const auto &hashes = updates.updatedFileHashes;
+    auto &changedFiles = updates.updatedFiles;
+    logger.debug("Trying to see if fast path is available after {} file changes", changedFiles.size());
+
+    ENFORCE(changedFiles.size() == hashes.size());
+    int i = -1;
+    {
+        for (auto &f : changedFiles) {
+            ++i;
+            auto fref = gs.findFileByPath(f->path());
+            if (!fref.exists()) {
+                logger.debug("Taking slow path because {} is a new file", f->path());
+                prodCategoryCounterInc("lsp.slow_path_reason", "new_file");
+                return false;
+            } else {
+                auto &oldHash = findHash(fref.id(), globalStateHashes, overriddingStateHashes);
+                ENFORCE(oldHash.definitions.hierarchyHash != core::GlobalStateHash::HASH_STATE_NOT_COMPUTED);
+                if (hashes[i].definitions.hierarchyHash == core::GlobalStateHash::HASH_STATE_INVALID) {
+                    logger.debug("Taking slow path because {} has a syntax error", f->path());
+                    prodCategoryCounterInc("lsp.slow_path_reason", "syntax_error");
+                    return false;
+                } else if (hashes[i].definitions.hierarchyHash != core::GlobalStateHash::HASH_STATE_INVALID &&
+                           hashes[i].definitions.hierarchyHash != oldHash.definitions.hierarchyHash) {
+                    logger.debug("Taking slow path because {} has changed definitions", f->path());
+                    prodCategoryCounterInc("lsp.slow_path_reason", "changed_definition");
+                    return false;
+                }
+            }
+        }
+    }
+    logger.debug("Taking fast path");
+    return true;
+}
+
+UnorderedMap<int, core::FileHash> mergeEvictions(const UnorderedMap<int, core::FileHash> &olderEvictions,
```

Comment:
Merges file hash evictions from two edits. Needed to make fast path decisions in the past, e.g. `olderEvictions` will contain the hashes evicted by the running slow path and `newerEvictions` will contain the hashes evicted by the update we are currently processing.